File: 54. Spiral Matrix/solution.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> l = new ArrayList<Integer>(matrix.length * matrix[0].length);
        int round = 0;
        while (matrix.length > round * 2 && matrix[0].length > round * 2) {
            int r = round;
            int c = round;
            l.add(matrix[r][c]);
            while (++r < matrix[0].length - round) {
                l.add(matrix[c][r]);
            }
            r--;
            while (++c < matrix.length - round) {
                l.add(matrix[c][r]);
            }
            c--;
            if (c == round || r == round) break;
            while (--r >= round) {
                l.add(matrix[c][r]);
            }
            r++;
            while (--c > round) {
                l.add(matrix[c][r]);
            }
            round++;
        }
        return l;
    }
}

File: 54. Spiral Matrix/solution.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> ans;
        int i = 0, j = 0;
        int total = matrix.size() * matrix[0].size();
        ans.push_back(matrix[i][j]);
        matrix[i][j] = 101;
        while (ans.size() < total) {
            while (j < matrix[0].size() - 1 && matrix[i][j + 1] != 101) {
                ans.push_back(matrix[i][++j]);
                matrix[i][j] = 101;
            }
            while (i < matrix.size() - 1 && matrix[i + 1][j] != 101) {
                ans.push_back(matrix[++i][j]);
                matrix[i][j] = 101;
            }
            while (j > 0 && matrix[i][j - 1] != 101) {
                ans.push_back(matrix[i][--j]);
                matrix[i][j] = 101;
            }
            while (i > 0 && matrix[i - 1][j] != 101) {
                ans.push_back(matrix[--i][j]);
                matrix[i][j] = 101;
            }
        }
        return ans;
    }
};